### PR TITLE
feat: add serialize/de-serialize

### PR DESCRIPTION
**Pull request by hansieodendaal**
- Added serialize and deserialize implementations
- Expanded the unit tests to test this
- Small refactoring

### CODE REVIEW

File: src/utils/generic.rs [hunk 1]
```diff
@@ -21,43 +24,56 @@ pub fn nonce(seed_nonce: &Scalar, label: &str, index: Option<usize>) -> Result<S
     let encoded_label = label.as_bytes();
     if encoded_label.len() > 16 {
         // See https://www.blake2.net/blake2.pdf section 2.8
-        return Err(ProofError::InvalidLength("nonce label".to_string()));
-    };
-    let hasher = if let Some(salt) = index {
+        Err(ProofError::InvalidLength("nonce label".to_string()))
+    } else if let Some(salt) = index {
```
Comment (SWvheerden) [inline src/utils/generic.rs:28]: I prefer the old code here. 
We introduce more nesting by removing the return

File: src/utils/generic.rs [hunk 2]
```diff
@@ -21,43 +24,56 @@ pub fn nonce(seed_nonce: &Scalar, label: &str, index: Option<usize>) -> Result<S
...
+            encoded_label,
+        )))
+    }
 }
 
 /// Decompose a given value into a vector of scalars for the required bit length
 pub fn bit_vector_of_scalars(value: u64, bit_length: usize) -> Result<Vec<Scalar>, ProofError> {
     if !bit_length.is_power_of_two() || bit_length > RangeProof::MAX_BIT_LENGTH {
-        return Err(ProofError::InvalidLength(
+        Err(ProofError::InvalidLength(
```
Comment (SWvheerden) [inline src/utils/generic.rs:52]: Same here, nested code reads and follows more difficult than non nested code.

**Review by SWvheerden — COMMENTED** [review]
Looks good to me. 
I just dont like the additional nested if's, 
They are a bad pattern

File: src/utils/generic.rs (hunk 1 above)
Comment (hansieodendaal) [reply src/utils/generic.rs:28]: This is the way @DenisKolodin proposed to get rid of additional `return` statements.

File: src/utils/generic.rs (hunk 2 above)
Comment (hansieodendaal) [reply src/utils/generic.rs:52]: See above

File: src/utils/generic.rs (hunk 1 above)
Comment (hansieodendaal) [reply src/utils/generic.rs]: Reverted

File: src/utils/generic.rs (hunk 2 above)
Comment (hansieodendaal) [reply src/utils/generic.rs]: Reverted

File: src/range_proof.rs [hunk 3]
```diff
@@ -600,6 +601,72 @@ impl RangeProof {
...
+                "Serialized proof bytes must be a factor of 32".to_string(),
+            ));
+        }
+        let num_elements = slice.len() / 32;
+        if (num_elements - 6) % 2 != 0 {
+            return Err(ProofError::InvalidLength(
+                "Serialized proof has incorrect number of elements".to_string(),
+            ));
+        }
+        let n = (num_elements - 6) / 2;
```
Comment (sdbondi) [inline src/range_proof.rs:636]: For `slice.len() == 0` this wraps around to a very large number.

**Review by sdbondi — COMMENTED** [review]
Looks good, I think that this needs more test cases, byte de/serialisation in general should always have extensive tests and ideally fuzzing.

File: src/range_proof.rs (hunk 3 above)
Comment (hansieodendaal) [reply src/range_proof.rs:636]: Thanks, fixed and tested

**Review by sdbondi — APPROVED** [review]
Looks good!

**Pull request merged**